packages/babel-parser/src/plugins/jsx/index.js

```diff
@@ -190,39 +187,54 @@ export default (superClass: Class<Parser>): Class<Parser> =>
     }
 
     jsxReadEntity(): string {
-      let str = "";
-      let count = 0;
-      let entity;
-      let ch = this.input[this.state.pos];
-
       const startPos = ++this.state.pos;
-      while (this.state.pos < this.length && count++ < 10) {
-        ch = this.input[this.state.pos++];
-        if (ch === ";") {
-          if (str[0] === "#") {
-            if (str[1] === "x") {
-              str = str.substr(2);
-              if (HEX_NUMBER.test(str)) {
-                entity = String.fromCodePoint(parseInt(str, 16));
-              }
-            } else {
-              str = str.substr(1);
-              if (DECIMAL_NUMBER.test(str)) {
-                entity = String.fromCodePoint(parseInt(str, 10));
-              }
-            }
-          } else {
-            entity = XHTMLEntities[str];
+      if (this.codePointAtPos(this.state.pos) === charCodes.numberSign) {
+        ++this.state.pos;
+
+        let radix = 10;
+        if (this.codePointAtPos(this.state.pos) === charCodes.lowercaseX) {
+          radix = 16;
+          ++this.state.pos;
+        }
+
+        const codePoint = this.readInt(
+          radix,
+          /* len */ undefined,
+          /* forceLen */ false,
+          /* allowNumSeparator */ "bail",
+        );
+        if (
+          codePoint &&
```

Comment:
```suggestion
          codePoint !== null &&
```
`&#0000;` is valid html entity.

Reply:
This might be another one to spec in JSX btw, because there’s likely divergence between implementations. There are a bunch of different things not allowed by XML/HTML/markdown (such as `\0` or lone surrogates)